pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,122 @@
+package addon
+
+import (
+	"context"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	csrclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	"k8s.io/klog/v2"
+
+	addoninformerv1alpha1 "github.com/open-cluster-management/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "github.com/open-cluster-management/api/client/addon/listers/addon/v1alpha1"
+)
+
+// addOnRegistrationController monitors ManagedClusterAddOn crs on hub and starts addOn registration
+// according to the setting from cr annotations.
+type addOnRegistrationController struct {
+	clusterName    string
+	agentName      string
+	kubeconfigData []byte
+	kubeClient     kubernetes.Interface
+	hubAddOnLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer
+	hubCSRClient   csrclient.CertificateSigningRequestInterface
+	recorder       events.Recorder
+
+	// addOnCertificateManagers maps the addon name to addOnCertificateManager
+	addOnCertificateManagers map[string]*addOnCertificateManager
```

Comment:
why do we have another layer of manager instead of having all clientcertcontroller managed here.

I would suggest we have 
```go
addOnCertificateManagers  map[string][registrationConfig]struct{manager: *clientcertManager, stopFunc: context.CancelFunc}.
```

So in each reconcile, you are able to build  the current []registrationConfig. And compare with the addOnCertificateManagers[addonName]

I think it is fine that we do not share secret informer in an addon as the first step. We can consider having a per ns secret informer since in our case, multiple addons are installed in one ns.

Reply:
I have to use two maps below instead the one you suggested.
```golang
	// registrationConfigs maps the addon name to a slice of registrationConfigs
	addOnRegistrationConfigs map[string][]registrationConfig
	// stopRegistrationFuncs maps the key of a registrationConfig to its stopRegistrationFunc
	stopRegistrationFuncs map[string]context.CancelFunc
```
`registrationConfig` struct is a invalid type for map key because registrationConfig.Subject.Organization is an array, which is not comparable.